### ISSUE TITLE
(deploy/amplify) Deploy Support for AWS Amplify

### ISSUE THREAD

**Issue by tylersayshi** [wakujs/waku#986]
A vite plugin is necessary to configure the build process to build for AWS Amplify.

It would be nice to include docs similar to: https://github.com/dai-shi/waku/blob/main/docs/builder/aws-lambda.mdx#L1-L4 for amplify

Also, it could be interesting to investigate nitro for adding this support.

**Comment by dai-shi** [comment]
@rmarscher @pmelab thoughts?

**Comment by dai-shi** [comment]
> nitro for adding this support.

Is it related to AWS Amplify?

**Comment by pmelab** [comment]
> @rmarscher @pmelab thoughts?

Is there a reason you tagged me? Unfortunately I don't have experience with AWS Amplify.

**Comment by dai-shi** [comment]
> Is there a reason you tagged me?

I thought you are familiar with AWS Lambda deployment, and I didn't know if there was a difference between Amplify and Lambda. Sorry about that.

Wait, I was completely wrong. I should have tagged @aheissenberger instead.

**Comment by pmelab** [comment]
> Wait, I was completely wrong. I should have tagged @aheissenberger instead.

😂 I suspected something like that.

**Comment by rmarscher** [comment]
> A vite plugin is necessary to configure the build process to build for AWS Amplify.

@tylersayshi Do you have a link to the docs for using Vite with AWS Amplify?

**Comment by tylersayshi** [comment]
@rmarscher do you mean just deploying any vite app at all to aws amplify? 

I don't really have any details on this beyond the initial ask though. I made this issue in response to a user asking for aws amplify support in the waku discord channel.

**Comment by rmarscher** [comment]
> @rmarscher do you mean just deploying any vite app at all to aws amplify?
> 
> I don't really have any details on this beyond the initial ask though. I made this issue in response to a user asking for aws amplify support in the waku discord channel.

Yeah, sorry. I found the context on Discord. I thought you meant that Amplify requires using a specific Vite plugin when deploying to it.

The Nitro presets are a pretty good reference for where the static and function assets need to be moved to for deploying to AWS amplify - https://github.com/nitrojs/nitro/blob/v2/src/presets/aws-amplify/preset.ts
https://nitro.build/deploy/providers/aws-amplify
I think we agreed to do our own thing for now and not try to incorporate Nitro's presets/config into our build system since it seems to be coupled with the rest of their ecosystem.

I am finding that a lot of deploy scenarios require a separate "function" folder with the server code from the public assets folder. A manifest of the public asset files is also often necessary.

I'm personally a bit confused on the correct cache headers for some of the files - especially the dist/public/RSC files. It seems like there are a few urls that do not update when the content changes on deploy. So I think those need to have no-cache headers or have their cache invalidated on deploy or something like that. Cloudfront cache invalidations are actually pretty expensive if you do them a lot. 

I don't personally have a use case for Amplify so I can't volunteer to take on supporting a deploy plugin. I'm happily using SST for now.

**Comment by tylersayshi** [comment]
Seems safe for this issue to wait for a stronger use case to develop itself - would be a nice option if someone is new to waku and wanting to use amplify I think. Or would be nice to see 👍's on the issue description to see that this is something many people would use.

**Comment by aheissenberger** [comment]
@tseijp very cool, great work!

Based on your example code with hono and as you build the Waku app for node (not for AWS Lambda) the app is deployed as a AWS Lambda docker container and not to `AWS Lambda@Edge`. Compared to the other deployment options you add an extra layer (wrapper) which might impacts performance and costs as all request events are converted to a http request to port 3000 to deploy the original app without any changes to the serving layer. It is very hard to compare the costs between these deployment options but it look like this option is slightly more expensive if you expect higher traffic and there are some fixed costs you pay per construct.

I will have a look at it as it looks like a very simple solution and comes with CI/CDI out of the box

**Comment by hi-ogawa** [comment]
I made an example of integrating Nitro https://nitro.build/ with Waku https://github.com/hi-ogawa/waku-nitro-example. I haven't tried but building with `NITRO_PRESET=aws_amplify` should provide deploy-able output and then you can follow Nitro's documentation https://nitro.build/deploy/providers/aws-amplify. I would appreciate if anyone tests this out this setup and please feel to share feedback!
